networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/rpc/ResponseStreamImpl.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.jvmlibp2p.rpc;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import tech.pegasys.artemis.util.sos.SimpleOffsetSerializable;
+
+public class ResponseStreamImpl<O extends SimpleOffsetSerializable> implements ResponseStream<O> {
+
+  private final CompletableFuture<Void> completionFuture = new CompletableFuture<>();
```

Comment:
Feels like there should be a way to get the result into this future ... As @cemozerr - maybe we need to have classes that handle responses that return 0,1, or many objects?

Reply:
So there are two things going on here.
Firstly there are two types of events - receiving data and an end of stream indication.  This `completionFuture` resolves when the end of stream is reached.  There may have been none, one or many responses in the mean-time.  We could make it a `CompletableFuture<List<O>>` and it could then contain the result, but that would mean all requests would get a list in response even if they only expect a single response and all responses would have to be buffered in memory before they can start being processed. If we're requesting a lot of blocks that's likely to be an issue and we really need to be able to stream the results.

Secondly, to @cemozerr 's comment, we wind up needing the single `ResponseStream` class because the libp2p `ProtocolBinding` interface implemented by `RpcMessageHandler` has to return a `CompletableFuture` of the `Controller` and the `Controller` has to define a single type of response it will return.  So currently the "response" it returns is the `ResponseStream` and it's returned as soon as the request is made (before any responses), then `ResponseStream` provides the logic to differentiate between 0, 1 or many responses. As above, we could just return a `List` for all requests to avoid this, but then we'd lose the ability to stream response data.

I've had a second look at this to see if there is a better way and struck out again. We could maybe hide it a little better by merging `ResponseStream` inside `RpcMessageHandler` but that class is already quite big and confusing so I think it's better to have the roles separated and then provide the nice facade to callers via `Peer`.

Reply:
> all responses would have to be buffered in memory before they can start being processed

very good point 👍